domain/cloudimagemetadata/service/service.go

```diff
@@ -0,0 +1,133 @@
+}
+
+// Service provides the API for working with cloud image metadata
+type Service struct {
+	st State
+}
+
+// NewService returns a new service around the persistence layer
+func NewService(st State) *Service {
+	return &Service{
+		st: st,
+	}
+}
+
+// SaveMetadata saves the provided cloud image metadata if non-empty and valid.
+// It returns a [errors.NothingToSave] if the input are empty, or a [errors.NotValid] if at least one of the inputs are
+// invalid.
+func (s Service) SaveMetadata(ctx context.Context, metadata []cloudimagemetadata.Metadata) error {
+	if len(metadata) == 0 {
+		return nil
+	}
+
+	if err := s.validateAllMetadata(metadata); err != nil {
+		return err
+	}
+
+	return s.st.SaveMetadata(ctx, metadata)
+}
+
+// DeleteMetadataWithImageID removes all the metadata associated with the given imageID from the state.
+func (s Service) DeleteMetadataWithImageID(ctx context.Context, imageId string) error {
+	if imageId == "" {
+		return errors.EmptyImageID
+	}
+
+	return s.st.DeleteMetadataWithImageID(ctx, imageId)
+}
+
+// FindMetadata retrieves a map of image metadata grouped by the source based on the provided filter criteria.
+func (s Service) FindMetadata(ctx context.Context, criteria cloudimagemetadata.MetadataFilter) (map[string][]cloudimagemetadata.Metadata, error) {
+	metadata, err := s.st.FindMetadata(ctx, criteria)
+	if err != nil {
+		return nil, err
+	}
+	metadataBySource := make(map[string][]cloudimagemetadata.Metadata)
+	for _, v := range metadata {
+		metadataBySource[v.Source] = append(metadataBySource[v.Source], v)
+	}
+	return metadataBySource, nil
+}
+
+// AllCloudImageMetadata retrieves all cloud image metadata from the state and returns them as a list.
+func (s Service) AllCloudImageMetadata(ctx context.Context) ([]cloudimagemetadata.Metadata, error) {
+	return s.st.AllCloudImageMetadata(ctx)
+}
+
+// validateAllMetadata validates a list of cloud image metadata.
+// It returns an error if any of the metadata entries are invalid.
+func (s Service) validateAllMetadata(metadata []cloudimagemetadata.Metadata) error {
+	var errs []error
+	for _, m := range metadata {
+		errs = append(errs, s.validateMetadata(m))
+	}
+	return jujuerrors.Join(errs...)
+}
+
+// validateMetadata validates a single cloud image metadata entry.
+func (s Service) validateMetadata(m cloudimagemetadata.Metadata) error {
+	if m.ImageID == "" {
+		return jujuerrors.Errorf("%w: %w", errors.EmptyImageID, errors.NotValid)
+	}
+
+	var missing []string
+	if m.Version == "" {
+		missing = append(missing, "version")
+	}
+	if m.Stream == "" {
+		missing = append(missing, "stream")
+	}
+	if m.Source == "" {
+		missing = append(missing, "source")
+	}
+	if m.Arch == "" {
+		missing = append(missing, "arch")
+	}
+	if m.Region == "" {
+		missing = append(missing, "region")
+	}
+
+	if len(missing) > 0 {
+		return errors.NotValidMissingFields(m.ImageID, missing)
+	}
+
+	supportedArchitectures := s.st.SupportedArchitectures(context.Background())
```

Comment:
You need to pass the context in here as a dependency. We have the unfortunate issue where the underlying database can sometimes lock up if we don't ensure the right context is passed through all the layers.

Reply:
Oh, nice spot. Done.